Current Map Feature Behind Feature Flag

AmiyaSX opened the pull request:
We need to merge the map feature into main before it goes too far.

Current functionalities:
- Basic graphic visuals, fetch booths data
- Basic interaction(switch floors, select booth/buildings)
- Basic filtering function on the sidebar
- Zoom in/out and auto focus on fair buildings
- DetailView on the sidebar of the company
- Edit mode for the map elements
- Put map feature behind feature flag
- Ability to set coordinates in url parameters

![image](https://github.com/user-attachments/assets/8167a899-7e16-4b80-b262-9d96b9c628b5)

vmork committed:
create map page, copy over demo code

vmork committed:
create map page, copy over demo code

vmork committed:
add types.ts, refactor, fetch exhibitors

vmork committed:
add types.ts, refactor, fetch exhibitors

vmork committed:
scale booth markers on zoom
color booths differently on hover

vmork committed:
scale booth markers on zoom
color booths differently on hover

vmork committed:
start adding locations

vmork committed:
start adding locations

vmork committed:
add sidebar to map view

vmork committed:
add sidebar to map view

vmork committed:
fly to new location center on change

vmork committed:
fly to new location center on change

vmork committed:
change sidebar ui on desktop

Was using the Drawer component for the sidebar for both small and large screens, but didnt work that well on desktop
and covered parts of map since it has to be positioned absolutely. Switched to just a div that can have
its width toggled to 0 with a button. Map width grows to fill remaining space

vmork committed:
change sidebar ui on desktop

Was using the Drawer component for the sidebar for both small and large screens, but didnt work that well on desktop
and covered parts of map since it has to be positioned absolutely. Switched to just a div that can have
its width toggled to 0 with a button. Map width grows to fill remaining space

vmork committed:
add detailed exhibitor view to sidebar

vmork committed:
add detailed exhibitor view to sidebar

AmiyaSX committed:
implement filtering search for sidebar

AmiyaSX committed:
implement filtering search for sidebar

vmork committed:
improve sidebar

vmork committed:
improve sidebar

vmork committed:
wip

vmork committed:
wip

vmork committed:
Merge branch 'main' into feat/map

AmiyaSX committed:
link interaction between sidebar and map

AmiyaSX committed:
link interaction between sidebar and map

vmork committed:
add filters to sidebar

vmork committed:
add filters to sidebar

vmork committed:
start working on edit mode

vmork committed:
start working on edit mode

vmork committed:
wip

vmork committed:
wip

vmork committed:
Merge remote-tracking branch 'origin/feat/map' into feat/map

vmork committed:
add ability to copy current geojson in editor

vmork committed:
add ability to copy current geojson in editor

vmork committed:
add highlight around buildings (nymble/library)

vmork committed:
add highlight around buildings (nymble/library)

vmork committed:
add ability to edit buildings

vmork committed:
add ability to edit buildings

vmork committed:
put map page behind feature flag

vmork committed:
put map page behind feature flag

vmork committed:
add toggle for edit mode, disabled in prod

vmork committed:
add toggle for edit mode, disabled in prod

AmiyaSX committed:
try fix lint check

AmiyaSX committed:
try fix lint check

AmiyaSX committed:
try fix format check

AmiyaSX committed:
try fix format check

AmiyaSX committed:
try fix format check again

AmiyaSX committed:
try fix lockfile

AmiyaSX committed:
try fix lockfile

vmork committed:
add ability to set coordinates in url parameters

vmork committed:
add ability to set coordinates in url parameters

AmiyaSX committed:
fix format checking

AmiyaSX committed:
Merge main branch

vmork committed:
Merge remote-tracking branch 'origin/feat/map' into feat/map

vmork committed:
style filtered booths differently

For now just decreasing the opacity of booths that don't match current filters,
but could easily tweak later

AmiyaSX committed:
Improve code structure

AmiyaSX committed:
Improve code structure

vmork committed:
extract filtering logic to new file

vmork committed:
Merge remote-tracking branch 'origin/feat/map' into feat/map

vmork committed:
attempt format fix

hampfh commented on package.json:
```diff
@@ -21,7 +22,11 @@
     "@tanstack/react-query": "^5.20.1",
+    "@turf/center-of-mass": "^7.0.0",
+    "@types/geojson": "^7946.0.14",
+    "@types/mapbox__mapbox-gl-draw": "^1.4.7",
```
Type packages should always be dev dependencies

vmork replied:
Moved to devDependencies

hampfh commented on src/app/globals.css:
```diff
@@ -1,3 +1,13 @@
 @tailwind components;
 @tailwind utilities;
+
+/* doesnt seem to work */
```
If it doesn't work then we should not add it, are you sure its not doing anything?

vmork replied:
That was for styling elements created by maplibre (the underlying map library used by react-map-gl) that can't be styled through react, like improving the look of the default popup and hiding the "attribution" thing that is always displayed in the bottom right corner of the map. But they don't seem to applied at all even though im pretty sure the selectors are correct, so thought it might be some tailwind-specific issue related to how global styles work (?), but couldn’t figure it out. Kept it because I was planning to come back to it later at some point

hampfh commented on src/app/student/_components/ExhibitorFilters.tsx:
```diff
@@ -0,0 +1,112 @@
+): FilterItem[] {
+  const distinct = new Map<FilterItem["id"], FilterItem>()
+  exhibitors.forEach(e => {
+    e[key].forEach(item => distinct.set(item.id, item))
```
Potential source for bug here, if you specify a field which isn't an array, you're going to iterate over undefined values. You have options, either, implement behavior that properly handles this, OR limit the keys that are available, for example only allow `employments` or `industries`

AmiyaSX replied:
I added an array check to make it safer. The key is defined at the beginning of the file and is limited to only `employments` or `industries`.

vmork replied:
Could be wrong but I don't think the runtime check is necessary since passing anything other than "employments" or "industries" will be a type error, and Exhibitor["employments" | "industries"] is known by typescript to be an array, so even adding non-array keys to the definition of FilterKey or changing the definition of Exhibitor would result in a type error in the filters file. So I ended up removing the check again, but let me know if you think this was incorrect

AmiyaSX replied:
I agree the checking might be not necessary here, since typescript already has the type check, and we already limited the key. Just to intentionally prevent the array to be null. But in the schema definition we don’t allow null value in Exhibitor object. So I think it is safe enough to not have the checking.

hampfh commented on src/app/student/_components/MultiSelect.tsx:
```diff
@@ -91,6 +88,7 @@ export default function MultiSelect({
       <PopoverContent
         className="z-10 w-full border-none p-0"
         align="start"
+        // side="bottom"
```
Why do we have a comment here?

vmork replied:
Removed, along with other uneccesary comments

hampfh commented on src/app/student/exhibitors/_components/ExhibitorListFilteringHeader.tsx:
This file doesn't seem to be used, is it possible that it is dead code?

vmork replied:
Yes, removed now

hampfh commented on src/app/student/lib/filters.ts:
What is the difference between these functions and the functions defined in `ExhibitorFilters.tsx`?

AmiyaSX replied:
No difference, I think it caused by some mistake while merging. Removed redundant code.

hampfh commented on src/app/student/map/_components/BoothPopup.tsx:
```diff
@@ -0,0 +1,19 @@
+
+export function BoothPopup({ booth }: { booth: Booth }) {
+  return (
+    <div>
```
What's the reason for this enclosing div?

vmork replied:
Removed

vmork replied:
Now that I think about it we should maybe just remove the popups completely, the only thing they display atm is the exhibitor name but that is not very useful since we already show detailed exhibitor information in the sidebar when you click on a booth, and not sure what else you could put in the popup. Thoughts?

hampfh replied:
Hi! Me and August discussed this slightly and I think that if we're keeping them then we should add more content to them (kind of like the popup in larm), if we don't want to add more content to them then we can remove the entire popup. I don't think that's a blocking thing for this PR though

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+}) {
+  // url: /student/map?location=[nymble/1|nymble/2|nymble/3|library]&lat=[number]&lng=[number]&zoom=[number]
+  // if location is not provided or is invalid, default to nymble/1
+  // if lat, lng or zoom is not provided, default to location center
```
Nice 👍

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+  const locationString = searchParams.get("location") ?? "nymble/1"
+  const locationIdString = locations.some(loc => loc.id === locationString)
+    ? locationString
+    : defaultLocation.id
```
Maybe this could be shortened a bit:

```ts
const location = locations.find(location => location.id === searchParams.get("location")) ?? defaultLocation.id
```

vmork replied:
That's a lot cleaner but I think we would need to change some other things for it to work. The issue is that `locationId` ("nymble/1" | "library" | etc) needs to be a state variable since it's updated by the location select component, and I dont think you can use `location` in the select component since it's an object and not a string. So we start by getting the locationId string and then lookup the location. Should maybe be restructured at some point to make it less confusing but couldn't figure out how for now

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+  const [locationId, setLocationId] = useState<LocationId>(
+    locationIdString as LocationId
+  )
+  const location = locations.find(loc => loc.id === locationId)!
```
unnecessary if fixing above comment

hampfh replied:
I would also suggest not using `!`, it might lead to trouble down the road since we're lying to typescript that we're sure the item exist

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+      </div>
+    )
+  }
+}
```
This could be moved into its own file

vmork replied:
Moved (and renamed to LocationSelect)

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+    : defaultLocation.id
+  const [locationId, setLocationId] = useState<LocationId>(
+    locationIdString as LocationId
+  )
```
I think locationId is a bit misleading, maybe call it "floor" instead

vmork replied:
Changed the url parameter to "floor" instead of "location" but ended up keeping location and locationId in the code. Agree that this terminology is somewhat confusing but it's already used in so many places in the code that it would be a bit difficult to change. Also not sure what to change it to, I feel like floor would make sense if we only had one building but the library isn't really a floor in my head

hampfh commented on src/app/student/map/_components/MapComponent.tsx:
```diff
@@ -0,0 +1,208 @@
+        }
+      }
+    }, [boothIds, stateKey])
+  }
```
Extract this hook, it might not be visible but nesting components, especially hooks, can create some nasty rerenders (and isn't super clear)

vmork replied:
Sorry did you mean extracting the entire useFeatureState function, or something else? Also just curious, how could it cause rerenders, isn't it just a function definition?

hampfh replied:
I think this video explains it quite well: https://www.youtube.com/watch?v=2sAdzy90GtE

But in essence, its problematic to have components inside other components (in this case a hook inside a component but it is the same thing) because it will cause the hook to re-render on every parent rerender

hampfh commented on src/app/student/map/_components/MapComponent.tsx:
```diff
@@ -0,0 +1,208 @@
+            id="cat"
+            type="raster"
+            ></Layer>
+        </Source> */}
```
Dead code

vmork replied:
Removed

hampfh commented on src/app/student/map/_components/MainView.tsx:
```diff
@@ -0,0 +1,134 @@
+      </div>
+    )
+  }
+}
```
Nested components

hampfh commented on src/app/student/map/editor/EditorMapComponent.tsx:
```diff
@@ -0,0 +1,372 @@
+    draw.set(currentFeatures)
+  }
+
+  useEffect(updateDrawFeatures, [location, mapLoaded])
```
Nit: I general I think its good to have all useEffects in one place to have better overview

hampfh commented on src/app/student/map/editor/EditorMapComponent.tsx:
```diff
@@ -0,0 +1,372 @@
+    setBooths(booths)
+  }
+
+  useEffect(updateBooths, [location, mapLoaded])
```
Here too

hampfh commented on src/app/student/map/editor/EditorMapComponent.tsx:
```diff
@@ -0,0 +1,372 @@
+  useEffect(updateBooths, [location, mapLoaded])
+
+  const activeFeature = activeFeatureId
+    ? featureMap.current.get(activeFeatureId)!
```
Unnecessary `!`

hampfh commented on src/app/student/map/editor/EditorMapComponent.tsx:
The components FeatureEditor and Toolbar are nested, please try to extract them

vmork replied:
Extracted to new files

vmork replied:
Also tried to clean up the editor code in general by moving things around and adding some comments, it's still pretty bad and needs to be refactored more at some point but just don't have time right now

hampfh commented on src/app/student/map/lib/locations.ts:
```diff
@@ -0,0 +1,44 @@
+  zoom: 18
+}
+
+export const locations: Location[] = [
```
maybe call this floors?

hampfh commented on src/app/student/map/page.tsx:
```diff
@@ -0,0 +1,51 @@
+    ])
+  )
+
+  const boothsByLocation: Map<LocationId, BoothMap> = new Map(
+    locations.map(loc => [loc.id, new Map()])
+  )
+  boothsById.forEach((booth, id) => {
+    boothsByLocation.get(booth.location)!.set(id, booth)
+  })
```
Maybe this can be memoized?

hampfh commented on src/app/student/map/page.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  return (
+    // TODO: pt-16 is to account for the navbar, will break if navbar size changes
+    <Suspense>
```
Since the entire page is in the suspense does it provide any value to have it? Maybe add a fallback, or extract parts of the page outside the suspense

AmiyaSX replied:
Yes the previous Vercel deployment error showed up when we didn't have it, that's why I added it here. It caused by calling useSearchParams() function in MainView.tsx, so basically the whole page is suspense, I could not come up a best extraction for only certain part of the page so far.

hampfh commented on src/feature_flags.ts:
```diff
@@ -7,14 +7,22 @@ export const FEATURE_FLAG_DEFINITIONS = {
 > = {
-  EVENT_PAGE: true
+  EVENT_PAGE: true,
+  MAP_PAGE: true
```
this should be `false` before we are completely done with the feature

vmork replied:
Fixed

hampfh commented on tailwind.config.ts:
```diff
@@ -52,11 +54,13 @@ const config = {
   },
-  // eslint-disable-next-line @typescript-eslint/no-require-imports
-  plugins: [require("tailwindcss-animate")]
+  plugins: [tailwindcssAnimate, containerQueries]
```
Thank you for fixing this 🙏 We want to get rid of commonjs

hampfh reviewed (commented):
Great work guys! Me and August has looked through the code, in general the feature is quite complete I think! The editor for the map seems to work really well and will be of great value for the person inputting the companies. The structure and logic of the map feels good! Nice!

There are quite a lot of smaller things through that needs to be addressed, not all of them necessarily has to be fixed now

Some general feedback:
* Try not to nest components, that will create problems down the road with renders etc
* There seem to be some dead code (out-commented code, unused files or just duplicated logic). Best practice would be to remove any code that is unused.
* Thank you for doing code documentation for parts of the code that might be of a bit higher complexity. It helps everybody understand the code a lot more!

vmork committed:
move type imports to devDependendcies

vmork committed:
cleanup and remove dead code

- removed ExhibitorListFilteringHeader which was not being used anymore
- moved ExhibitorFilters from student/components -> student/exhibitors/components since its not used outside exhibitors/
- renamed 'MapListFilteringHeader' -> 'MapFilters' to match 'ExhibitorFilters'
- removed map/components/Layers.tsx (code was already moved to map/lib/config.ts)

vmork committed:
remove unneccesary div in BoothPopup

vmork committed:
refactor handling of invalid location

vmork committed:
move LocationSelect to new file

AmiyaSX committed:
Fix filter code duplicate

AmiyaSX committed:
Make Suspense only wraps MainView

AmiyaSX committed:
iterate array safely

vmork committed:
editor wip

vmork committed:
Merge remote-tracking branch 'origin/feat/map' into feat/map

vmork committed:
set MAP_PAGE feature flag to false

hampfh commented:
Great progress, I think we're close to being able to merge this, the blocking thing for me is just the nested components now. If all of those are addressed we can fix the rest at a later point

vmork committed:
extract useFeatureState function to outside component

hampfh reviewed (approved):
Great! I think this is good for now, we can continue to iterate after this is merged
